### PR TITLE
Add support for global advisories and unify some shared logic with repository advisories

### PR DESCRIPTION
**Pull request by crimsonknave**
This PR adds support for the new Global Advisories api: https://docs.github.com/en/rest/security-advisories/global-advisories. It adds the `get_global_advisory` and `get_global_advisories` methods.

It creates a new `GlobalAdvisory` class  and pulls the shared fields into a class both Global and Repository Advisories inherit from. The payloads of the Global and Repository Advisories are similar, but not quite the same. Repository Advisories have fields designed to allow users to submit using the payloads that are returned. Global Advisories have additional fields that do not apply to advisories attached to a repository. I also implemented a few more fields that are present in both, but were not included (CVSS and identifiers).

I have attempted to follow the code style in the other files, happy to update anything where I'm not following convention. 

-----

The fact that I renamed some of the files means that the header script is removing the copyright from anyone who was listed in the original file, which seems wrong. But, I'm not sure what the right fix is, any manual editing would be overwritten by the next person to run the script.

Side note: `script/fix_headers.py` adds `# -*- coding: utf-8 -*-` which is then removed by the `pyupgrade` step of the pre-commit hooks.

### CODE REVIEW

File: tests/Github_.py
Comment (EnricoMi) [inline tests/Github_.py:3]: removing these lines does not look right

File: github/AdvisoryCredit.py
Comment (EnricoMi) [inline github/AdvisoryCredit.py:3]: removing this line does not look right either

File: github/CVSS.py
Comment (EnricoMi) [inline github/CVSS.py:40]: Why not float?

**Review by EnricoMi — CHANGES_REQUESTED** [review]

File: tests/Github_.py
Comment (crimsonknave) [reply tests/Github_.py:3]: Yup, that was done by the automated script. I can manually fix it, but as I noted in the description anyone that runs the script will just revert it. It looks like this file was renamed a few years ago and the script does not follow the rename backwards, so it only sees anyone who touched the new file.

File: github/AdvisoryCredit.py
Comment (crimsonknave) [reply github/AdvisoryCredit.py:3]: Same issue here, this file was renamed so the script removed it.

File: github/CVSS.py
Comment (crimsonknave) [reply github/CVSS.py:40]: Fundamentally the score is a user facing number so having it be a decimal means that it will always look nice. There won't be a CVSS score that needs more than 1-2 decimal places, so we don't need any of the float precision.

Additionally, decimal will ensure that we don't end up with 5.00000000000001 (or whatever) when doing any math or converting with a score. It also means I can be confident that any comparisons I do against the score won't be impacted by the same issue (so, if I check if the score is 5 less for example, the above float value.

The [decimal docs](https://docs.python.org/3/library/decimal.html) go into more detail into the pros and cons.

**Comment by crimsonknave** [comment]
Happy to make whatever changes y'all would like. I just wasn't sure what the right answer was around the comments.

**Comment by crimsonknave** [comment]
@EnricoMi I'm happy to resolve the copyright comments however you think is best. Aside from that and the merge conflicts, is there anything else I should do to get this PR ready to merge?

File: tests/Github_.py
Comment (EnricoMi) [reply tests/Github_.py:3]: Fixed in #2728, you can ignore this.

File: github/CVSS.py
Comment (EnricoMi) [reply github/CVSS.py:41]: Sounds good, thanks!

**Comment by EnricoMi** [comment]
@JLLeitschuh are you happy with this refactoring of your advisories?

**Comment by EnricoMi** [comment]
@crimsonknave I am happy with this work, please resolve the conflicts and we are ready to go ahead (pending @JLLeitschuh approval).

File: tests/Github_.py
Comment (crimsonknave) [reply tests/Github_.py:3]: Ignore this as in "these changes are fine to commit as is" or, "revert and just add myself and they'll be updated once that PR is merged"?

Comment (EnricoMi) [reply tests/Github_.py:3]: Fine to ignore, they will be restored in #2801.

**Review by JLLeitschuh — APPROVED** [review]
This looks good to me! Happy to merge!

**Comment by JLLeitschuh** [comment]
@EnricoMi can I merge this?

**Comment by JLLeitschuh** [comment]
@EnricoMi friendly ping!

**Review by EnricoMi — APPROVED** [review]
LGTM!

**Pull request merged**
